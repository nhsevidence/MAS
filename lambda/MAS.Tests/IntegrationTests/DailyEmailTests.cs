using MailChimp.Net.Core;
using MailChimp.Net.Interfaces;
using MailChimp.Net.Models;
using MAS.Configuration;
using MAS.Models;
using MAS.Services;
using MAS.Tests.Infrastructure;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Mvc.Controllers;
using Microsoft.AspNetCore.Mvc.Testing;
using Microsoft.AspNetCore.Mvc.ViewFeatures;
using Microsoft.AspNetCore.Routing;
using Microsoft.AspNetCore.TestHost;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyInjection.Extensions;
using Microsoft.Extensions.Logging;
using Moq;
using Shouldly;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System.Web.Mvc;
using Xunit;
using Source = MAS.Models.Source;

namespace MAS.Tests.IntegrationTests
{
    public class TestMailService : IMailService
    {
        public Task<string> CreateAndSendDailyAsync(string subject, string previewText, string body, List<string> specialitiesInEmail)
        {
            throw new NotImplementedException();
        }
    }

    public class DailyEmailTests : TestBase
    {
        [Fact]
        public async void EmailBodyHtmlMatchesApproved()
        {
            // Arrange
            var fakeMailService = new Mock<IMailService>();

            string bodyHtml = string.Empty;
            fakeMailService.Setup(s => s.CreateAndSendDailyAsync(It.IsAny<string>(), It.IsAny<string>(), It.IsAny<string>(), It.IsAny<List<string>>()))
                .Callback<string, string, string, List<string>>((subject, previewText, body, specialitiesInEmail) => bodyHtml = body)
                .ReturnsAsync("1234");

            var client = WithImplementation(fakeMailService.Object).CreateClient();

            // Act
            var response = await client.PutAsync("/api/mail/daily", null);

            // Assert
            response.StatusCode.ShouldBe(System.Net.HttpStatusCode.OK);

            var responseText = await response.Content.ReadAsStringAsync();
            responseText.ShouldBe("1234");

            bodyHtml.ShouldMatchApproved();
        }

        [Fact]
        public async void EmailBodyHtmlMatchesApprovedForSingleItem()
        {
            // Arrange
            var fakeMailService = new Mock<IMailService>();

            string bodyHtml = string.Empty;
            fakeMailService.Setup(s => s.CreateAndSendDailyAsync(It.IsAny<string>(), It.IsAny<string>(), It.IsAny<string>(), It.IsAny<List<string>>()))
                .Callback<string, string, string, List<string>>((subject, previewText, body, specialitiesInEmail) => bodyHtml = body)
                .ReturnsAsync("1234");

            var client = WithImplementation(fakeMailService.Object).CreateClient();

            AppSettings.CMSConfig.DailyItemsPath = "/daily-items-single.json";

            // Act
            var response = await client.PutAsync("/api/mail/daily", null);

            // Assert
            response.StatusCode.ShouldBe(System.Net.HttpStatusCode.OK);

            var responseText = await response.Content.ReadAsStringAsync();
            responseText.ShouldBe("1234");

            bodyHtml.ShouldMatchApproved();
        }

        [Fact]
        public async void EmailBodyHtmlMatchesApprovedForSingleItemWithSpecificDate()
        {
            // Arrange
            var fakeMailService = new Mock<IMailService>();

            string bodyHtml = string.Empty;
            fakeMailService.Setup(s => s.CreateAndSendDailyAsync(It.IsAny<string>(), It.IsAny<string>(), It.IsAny<string>(), It.IsAny<List<string>>()))
                .Callback<string, string, string, List<string>>((subject, previewText, body, specialitiesInEmail) => bodyHtml = body)
                .ReturnsAsync("1234");

            var client = WithImplementation(fakeMailService.Object).CreateClient();

            AppSettings.CMSConfig.DailyItemsPath = "/daily-items-{0}.json";

            // Act
            var response = await client.PutAsync("/api/mail/daily?date=01-01-2020", null);

            // Assert
            response.StatusCode.ShouldBe(System.Net.HttpStatusCode.OK);

            var responseText = await response.Content.ReadAsStringAsync();
            responseText.ShouldBe("1234");

            bodyHtml.ShouldMatchApproved();
        }

    }
}
