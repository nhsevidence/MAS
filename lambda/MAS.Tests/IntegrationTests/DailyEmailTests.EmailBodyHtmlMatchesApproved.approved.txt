


<table class="row article-border article-mec article-section-header">
    <tbody>
        <tr>
            <th class="small-12 large-12 first last">
                <table>
                    <tbody>
                        <tr>
                            <th>
                                <p>
                                    *|INTERESTED:Daily specialities of interest:Anaesthesia and pain,Complementary and alternative therapies|*
                                        Articles that match your chosen specialities today
                                    *|ELSE:|*
                                        No articles match your chosen specialities today
                                    *|END:INTERESTED|*
                                    - <a class="subtle-link" href="*|UPDATE_PROFILE|*">Edit your subscription</a>
                                </p>
                            </th>
                            <th class="expander"></th>
                        </tr>
                    </tbody>
                </table>
            </th>
        </tr>
    </tbody>
</table>


    *|INTERESTED:Daily specialities of interest:Anaesthesia and pain,Complementary and alternative therapies|*

<table class="spacer-heading">
    <tbody>
        <tr>
            <td height="16px" style="font-size:16px; line-height:16px;">
                <h3 class="article-heading">
                    Evidence summary
                </h3>
            </td>
        </tr>
    </tbody>
</table>        *|INTERESTED:Daily specialities of interest:Anaesthesia and pain,Complementary and alternative therapies|*

<table class="row article-border">
    <tbody>
        <tr>
            <th class="small-12 large-12 columns first last">
                <table>
                    <tbody>
                        <tr>
                            <th>
                                <h4 class="h4 article-title">
                                    <a class="article-link" href="https://www.nhs.uk/">
                                        Wonder Drug
                                    </a>
                                </h4>
                                <p class="article-description">
                                    testing
                                </p>
                                    <p class="article-description">
                                        <a class="article-link-subtle standard-link" href="http://mas-integration-tests/testing-789.html" aria-label="Read SPS commentary for Wonder Drug">
                                            Read SPS commentary
                                        </a>
                                    </p>
                                <p class="article-tags">
                                    <span class="source-tag">
                                        ACP Journal Club
                                    </span>
                                        <span class="specialism-tag">
                                            Anaesthesia and pain
                                        </span>
                                        <span class="specialism-tag">
                                            Complementary and alternative therapies
                                        </span>
                                </p>
                            </th>
                            <th class="expander"></th>
                        </tr>
                    </tbody>
                </table>
            </th>
        </tr>
    </tbody>
</table>        *|END:INTERESTED|*
        *|INTERESTED:Daily specialities of interest:Anaesthesia and pain,Complementary and alternative therapies|*

<table class="row article-border">
    <tbody>
        <tr>
            <th class="small-12 large-12 columns first last">
                <table>
                    <tbody>
                        <tr>
                            <th>
                                <h4 class="h4 article-title">
                                    <a class="article-link" href="https://www.bbc.co.uk/">
                                        Hello world
                                    </a>
                                </h4>
                                <p class="article-description">
                                    test
                                </p>
                                    <p class="article-description">
                                        <a class="article-link-subtle standard-link" href="http://mas-integration-tests/hello-world.html" aria-label="Read SPS commentary for Hello world">
                                            Read SPS commentary
                                        </a>
                                    </p>
                                <p class="article-tags">
                                    <span class="source-tag">
                                        Substance Misuse Management in General Practice
                                    </span>
                                        <span class="specialism-tag">
                                            Anaesthesia and pain
                                        </span>
                                        <span class="specialism-tag">
                                            Complementary and alternative therapies
                                        </span>
                                </p>
                            </th>
                            <th class="expander"></th>
                        </tr>
                    </tbody>
                </table>
            </th>
        </tr>
    </tbody>
</table>        *|END:INTERESTED|*
    *|END:INTERESTED|*
    *|INTERESTED:Daily specialities of interest:Anaesthesia and pain,Complementary and alternative therapies|*

<table class="spacer-heading">
    <tbody>
        <tr>
            <td height="16px" style="font-size:16px; line-height:16px;">
                <h3 class="article-heading">
                    Evidence summary 2
                </h3>
            </td>
        </tr>
    </tbody>
</table>        *|INTERESTED:Daily specialities of interest:Anaesthesia and pain,Complementary and alternative therapies|*

<table class="row article-border">
    <tbody>
        <tr>
            <th class="small-12 large-12 columns first last">
                <table>
                    <tbody>
                        <tr>
                            <th>
                                <h4 class="h4 article-title">
                                    <a class="article-link" href="https://www.nice.org.uk/">
                                        MY MEDICINE
                                    </a>
                                </h4>
                                <p class="article-description">
                                    this is my medicine
                                </p>
                                    <p class="article-description">
                                        <a class="article-link-subtle standard-link" href="http://mas-integration-tests/my-medicine.html" aria-label="Read SPS commentary for MY MEDICINE">
                                            Read SPS commentary
                                        </a>
                                    </p>
                                <p class="article-tags">
                                    <span class="source-tag">
                                        ADHD Awareness Month
                                    </span>
                                        <span class="specialism-tag">
                                            Anaesthesia and pain
                                        </span>
                                        <span class="specialism-tag">
                                            Complementary and alternative therapies
                                        </span>
                                </p>
                            </th>
                            <th class="expander"></th>
                        </tr>
                    </tbody>
                </table>
            </th>
        </tr>
    </tbody>
</table>        *|END:INTERESTED|*
        *|INTERESTED:Daily specialities of interest:Anaesthesia and pain,Complementary and alternative therapies|*

<table class="row article-border">
    <tbody>
        <tr>
            <th class="small-12 large-12 columns first last">
                <table>
                    <tbody>
                        <tr>
                            <th>
                                <h4 class="h4 article-title">
                                    <a class="article-link" href="https://www.google.com/">
                                        A Placebo
                                    </a>
                                </h4>
                                <p class="article-description">
                                    hello
                                </p>
                                    <p class="article-description">
                                        <a class="article-link-subtle standard-link" href="http://mas-integration-tests/hello-there.html" aria-label="Read SPS commentary for A Placebo">
                                            Read SPS commentary
                                        </a>
                                    </p>
                                <p class="article-tags">
                                    <span class="source-tag">
                                        Journal of Surgical Oncology
                                    </span>
                                        <span class="specialism-tag">
                                            Anaesthesia and pain
                                        </span>
                                        <span class="specialism-tag">
                                            Complementary and alternative therapies
                                        </span>
                                </p>
                            </th>
                            <th class="expander"></th>
                        </tr>
                    </tbody>
                </table>
            </th>
        </tr>
    </tbody>
</table>        *|END:INTERESTED|*
    *|END:INTERESTED|*


*|INTERESTED:Send me everything from Medicines Awareness Daily:Send me everything|*
<table class="row article-border article-mec article-section-header">
    <tbody>
        <tr>
            <th class="small-12 large-12 first last">
                <table>
                    <tbody>
                        <tr>
                            <th>
                                <p>
                                    All articles from Medicines awareness services: Daily edition
                                </p>
                            </th>
                            <th class="expander"></th>
                        </tr>
                    </tbody>
                </table>
            </th>
        </tr>
    </tbody>
</table>


<table class="spacer-heading">
    <tbody>
        <tr>
            <td height="16px" style="font-size:16px; line-height:16px;">
                <h3 class="article-heading">
                    Evidence summary
                </h3>
            </td>
        </tr>
    </tbody>
</table>
<table class="row article-border">
    <tbody>
        <tr>
            <th class="small-12 large-12 columns first last">
                <table>
                    <tbody>
                        <tr>
                            <th>
                                <h4 class="h4 article-title">
                                    <a class="article-link" href="https://www.nhs.uk/">
                                        Wonder Drug
                                    </a>
                                </h4>
                                <p class="article-description">
                                    testing
                                </p>
                                    <p class="article-description">
                                        <a class="article-link-subtle standard-link" href="http://mas-integration-tests/testing-789.html" aria-label="Read SPS commentary for Wonder Drug">
                                            Read SPS commentary
                                        </a>
                                    </p>
                                <p class="article-tags">
                                    <span class="source-tag">
                                        ACP Journal Club
                                    </span>
                                        <span class="specialism-tag">
                                            Anaesthesia and pain
                                        </span>
                                        <span class="specialism-tag">
                                            Complementary and alternative therapies
                                        </span>
                                </p>
                            </th>
                            <th class="expander"></th>
                        </tr>
                    </tbody>
                </table>
            </th>
        </tr>
    </tbody>
</table>
<table class="row article-border">
    <tbody>
        <tr>
            <th class="small-12 large-12 columns first last">
                <table>
                    <tbody>
                        <tr>
                            <th>
                                <h4 class="h4 article-title">
                                    <a class="article-link" href="https://www.bbc.co.uk/">
                                        Hello world
                                    </a>
                                </h4>
                                <p class="article-description">
                                    test
                                </p>
                                    <p class="article-description">
                                        <a class="article-link-subtle standard-link" href="http://mas-integration-tests/hello-world.html" aria-label="Read SPS commentary for Hello world">
                                            Read SPS commentary
                                        </a>
                                    </p>
                                <p class="article-tags">
                                    <span class="source-tag">
                                        Substance Misuse Management in General Practice
                                    </span>
                                        <span class="specialism-tag">
                                            Anaesthesia and pain
                                        </span>
                                        <span class="specialism-tag">
                                            Complementary and alternative therapies
                                        </span>
                                </p>
                            </th>
                            <th class="expander"></th>
                        </tr>
                    </tbody>
                </table>
            </th>
        </tr>
    </tbody>
</table>
<table class="spacer-heading">
    <tbody>
        <tr>
            <td height="16px" style="font-size:16px; line-height:16px;">
                <h3 class="article-heading">
                    Evidence summary 2
                </h3>
            </td>
        </tr>
    </tbody>
</table>
<table class="row article-border">
    <tbody>
        <tr>
            <th class="small-12 large-12 columns first last">
                <table>
                    <tbody>
                        <tr>
                            <th>
                                <h4 class="h4 article-title">
                                    <a class="article-link" href="https://www.nice.org.uk/">
                                        MY MEDICINE
                                    </a>
                                </h4>
                                <p class="article-description">
                                    this is my medicine
                                </p>
                                    <p class="article-description">
                                        <a class="article-link-subtle standard-link" href="http://mas-integration-tests/my-medicine.html" aria-label="Read SPS commentary for MY MEDICINE">
                                            Read SPS commentary
                                        </a>
                                    </p>
                                <p class="article-tags">
                                    <span class="source-tag">
                                        ADHD Awareness Month
                                    </span>
                                        <span class="specialism-tag">
                                            Anaesthesia and pain
                                        </span>
                                        <span class="specialism-tag">
                                            Complementary and alternative therapies
                                        </span>
                                </p>
                            </th>
                            <th class="expander"></th>
                        </tr>
                    </tbody>
                </table>
            </th>
        </tr>
    </tbody>
</table>
<table class="row article-border">
    <tbody>
        <tr>
            <th class="small-12 large-12 columns first last">
                <table>
                    <tbody>
                        <tr>
                            <th>
                                <h4 class="h4 article-title">
                                    <a class="article-link" href="https://www.google.com/">
                                        A Placebo
                                    </a>
                                </h4>
                                <p class="article-description">
                                    hello
                                </p>
                                    <p class="article-description">
                                        <a class="article-link-subtle standard-link" href="http://mas-integration-tests/hello-there.html" aria-label="Read SPS commentary for A Placebo">
                                            Read SPS commentary
                                        </a>
                                    </p>
                                <p class="article-tags">
                                    <span class="source-tag">
                                        Journal of Surgical Oncology
                                    </span>
                                        <span class="specialism-tag">
                                            Anaesthesia and pain
                                        </span>
                                        <span class="specialism-tag">
                                            Complementary and alternative therapies
                                        </span>
                                </p>
                            </th>
                            <th class="expander"></th>
                        </tr>
                    </tbody>
                </table>
            </th>
        </tr>
    </tbody>
</table>*|END:INTERESTED|*
